### PR TITLE
REVERTED: Fix stackoverflow when referenced by self annotation

### LINKED ISSUE #11870: 2.13.1 scalac StackOverflowError when annotation argument references the annottee
Consider the example:
```scala

class Invariant(condition: Unit => Boolean) extends scala.annotation.StaticAnnotation
abstract class Demo {
  @Invariant( _ => x.length == 3 )
  def x: String = "foo"
}
```

This code results in the `java.lang.StackOverflowError` through `scala.reflect.internal.AnnotationInfos$LazyAnnotationInfo.forcedInfo$lzycompute(AnnotationInfos.scala:158)` ([scalac-SOE-stack.txt](https://github.com/scala/bug/files/4129572/scalac-SOE-stack.txt) has the full stack trace if you need it). As far as I can tell, the stack trace is different from those encountered in #7449, #1632, #1667 and #10421. Increasing the stack `-Xss128m` just makes the SOE stack trace longer, so it looks like cyclic init.

Some pointers to help narrow the problem.

1. The annotation argument does not have to be a lambda. The following snippet yields the same SOE: 
```scala
class Annot(arg: Any) extends scala.annotation.StaticAnnotation
abstract class Demo {
  @Annot(x)
  def x: String = "foo"
}
```

2. The following example compiles fine.
```scala
class Annot(arg: Any) extends scala.annotation.StaticAnnotation
abstract class Demo {
  @Annot(y)
  def x: String = "foo"
  def y: String = "bar"
}
```

3. However, if you put the annotation on the second field as well, there is the SOE again
```scala
class Annot(arg: Any) extends scala.annotation.StaticAnnotation
abstract class Demo {
  @Annot(y)
  def x: String = "foo"

 @Annot(x)
  def y: String = "bar"
}
```

4. Finally, the code below causes SOE even when the annotations are different.
```scala
class Annot1(arg: Any) extends scala.annotation.StaticAnnotation
class Annot2(arg: Any) extends scala.annotation.StaticAnnotation
abstract class Demo {
  @Annot1(y)
  def x: String = "foo"

  @Annot2(x)
  def y: String = "bar"
}
```

Thank you!

### PR DESCRIPTION
**Pull request by changvvb**
Fixes https://github.com/scala/bug/issues/11870

(reverted in https://github.com/scala/scala/pull/9069)

### CODE REVIEW

**Comment by changvvb** [comment]
Jenkins pipeline is pending at https://scala-ci.typesafe.com/job/scala-2.13.x-validate-main/12079/console because error happened. Maybe it should be closed manually.

**Comment by SethTisue** [comment]
@changvvb would you mind rebasing against the current head of 2.13.x? that should make the problem go away

**Comment by changvvb** [comment]
> @changvvb would you mind rebasing against the current head of 2.13.x? that should make the problem go away

@SethTisue It seems not work  😕.

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala [hunk 1]
```diff
@@ -1884,13 +1884,16 @@ trait Namers extends MethodSynthesis {
      */
     def annotSig(annotations: List[Tree], annotee: Tree, pred: AnnotationInfo => Boolean): List[AnnotationInfo] =
       annotations filterNot (_ eq null) map { ann =>
-        val ctx = typer.context
```
Comment (dwijnand) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:1887]: I'm guessing it's better to keep this out of the `lazily`.

Comment (changvvb) [reply src/compiler/scala/tools/nsc/typechecker/Namers.scala:1887]: OK, I realized typer.context is mutable.

**Comment by dwijnand** [comment]
I'll re-fix the commits so we can reduce the noise merging this PR would bring.

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala [hunk 2]
```diff
@@ -1885,12 +1885,18 @@ trait Namers extends MethodSynthesis {
...
-          }
+        // enteringTyper to allow inferView in annotation args, scala/bug#5892
+        def createAnnotationInfo = enteringTyper {
+          val annotSig = newTyper(ctx.makeNonSilent(ann)).typedAnnotation(ann, Some(annotee))
+          if (pred(annotSig)) annotSig else UnmappableAnnotation // UnmappableAnnotation will be dropped in typedValDef and typedDefDef
+        }
+
+        ann match {
+          case Apply(Select(New(Ident(name)),_),_) if name == uncheckedStableClass.name =>
+            createAnnotationInfo
```
Comment (lrytz) [inline src/compiler/scala/tools/nsc/typechecker/Namers.scala:1897]: I think this is too breaking. The following example compiles in 2.13.x, but not with this PR (we should add a test for that):

```scala
import scala.annotation.unchecked

class C {
  class D
}

class Test {
  locally {
    val c = new C
    import c._
    new D
  }

  locally {
    import unchecked.uncheckedStable
    @uncheckedStable def c = new C
    import c._ // ok
    new D
  }

  locally {
    @unchecked.uncheckedStable def c = new C
    import c._ // error, the annotation is not `New(Ident(name))`
    new D
  }
  
  locally {
    import unchecked.{uncheckedStable => uc}
    @uc def c = new C
    import c._ // error, the annotation is `New(Ident(name))`, but `name` is different
    new D
  }
}
```

Idea: we could probably compute all annotations eagerly that have no, or an empty argument list. That would include `uncheckedStable`.

**Review by lrytz — CHANGES_REQUESTED** [review]

File: src/compiler/scala/tools/nsc/typechecker/Namers.scala (hunk 2 above)
Comment (changvvb) [reply src/compiler/scala/tools/nsc/typechecker/Namers.scala:1897]: Great example and idea. I was careless about it.

**Review by lrytz — APPROVED** [review]
thanks, @changvvb

**Comment by SethTisue** [comment]
@lrytz over at https://github.com/scala/community-build/pull/1164 this caused a regression; one Akka source file started failing to compile. I isolated it to just:

```
import S._
@annotation.strictfp object S

error: encountered unrecoverable cycle resolving import.
Note: this is often due in part to a class depending on a definition nested within its companion.
If applicable, you may wish to try moving some members into another object.
```

(the exact annotation isn't important, e.g. `@nowarn` does the same thing)

such an import is a bit unusual. the Akka folks wanted it because they were doing this:

```
import S._
class C(x: Inner)
@internalApi object S { trait Inner}
```

and note that you can't move the import inside the class definition because the location where they want to refer to `Inner` is in a constructor parameter.

interestingly, Dotty has the same bug or limitation, whichever you want to call it:

```
scala> import S._ ; object S                                                                       
// defined object S

scala> import S._ ; @annotation.strictfp object S                                                                       
1 |import S._ ; @annotation.strictfp object S
  |              ^
  |              Cyclic reference involving module class S$
```

not sure if this is a 2.13.3-blocker or not. in the Akka source the problem is easy to work around because you can either 1) `import S.Inner` instead of `import S._`, or 2) skip the import entirely and just write out `x: S.Inner`

**Comment by SethTisue** [comment]
@smarter would you like me to open a lampepfl/dotty ticket on this?

**Comment by retronym** [comment]
I would suggest reverting this change for 2.13.3. Lazy typechecking of annotations can be relied on in pretty subtle ways.

Perhaps our annotation typechecking should have an extra phase:

   1. Lazy we know nothing about the annotation
   2. Typecheck the annotation qualifier, but not the arguments. This would be enough to answer `Symbol#hasAnnotation` / `AnnotationInfo.matches`, and would be transparently forced when calling these methods.
   3. Fully typechecked

If there are still cases that could run into a `StackOverflowError` in the `uncheckedStable` lookup, we could add cycle detection to turn this into a cyclic error report.

**Comment by lrytz** [comment]
I like the suggestion of lazily type checking the parts of an `AnnotationInfo` and making the whole thing more transparent.

I also agree that reverting is the safer choice for 2.13.3.

**Comment by changvvb** [comment]
Yes, Symbol#hasAnnotation can be more clever.

**Pull request merged**